exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -0,0 +1,245 @@
+    checkNotNull(nodeType);
+    return new Builder(nodeType);
+  }
+
+  /**
+   * Returns a list of user service adapters created from given service modules.
+   */
+  private List<UserServiceAdapter> toUserServiceAdapters(List<Class<? extends ServiceModule>> serviceModules) {
+    List<UserServiceAdapter> serviceAdapters = new ArrayList<>(serviceModules.size());
+    for (Class<? extends ServiceModule> moduleClass : serviceModules) {
+      UserServiceAdapter serviceAdapter = createUserServiceAdapter(moduleClass);
+      serviceAdapters.add(serviceAdapter);
+    }
+    return serviceAdapters;
+  }
+
+  /**
+   * Populate the map of services given the list of service adapters.
+   */
+  private void populateServiceMap(List<UserServiceAdapter> serviceAdapters) {
+    for (UserServiceAdapter serviceAdapter: serviceAdapters) {
+      short serviceId = serviceAdapter.getId();
+      checkForDuplicateService(serviceId);
+      services.put(serviceId, serviceAdapter.getService());
+    }
+  }
+
+  private void checkForDuplicateService(short serviceId) {
+    checkArgument(!services.containsKey(serviceId),
+        "Service with id %s was added to the TestKit twice");
+  }
+
+  /**
+   * Builder for the TestKit.
+   */
+  public static final class Builder {
+
+    private EmulatedNodeType nodeType;
+    private short validatorCount;
+    private List<Class<? extends ServiceModule>> services = new ArrayList<>();
+    private TimeProvider timeProvider;
+
+    private Builder(EmulatedNodeType nodeType) {
+      // TestKit network should have at least one validator node
+      if (nodeType == EmulatedNodeType.AUDITOR) {
+        validatorCount = 1;
+      }
+      this.nodeType = nodeType;
+    }
+
+    /**
+     * Sets number of additional validator nodes in the TestKit network. Note that
+     * regardless of the configured number of validators, only a single service will be
+     * instantiated.
+     */
+    public Builder withValidators(short validatorCount) {
+      this.validatorCount = validatorCount;
+      return this;
+    }
+
+    /**
+     * Adds a service with which the TestKit would be instantiated. Several services can be added.
+     */
+    public Builder withService(Class<? extends ServiceModule> serviceModule) {
+      services.add(serviceModule);
+      return this;
+    }
+
+    /**
+     * Adds a variable number of services with which the TestKit would be instantiated.
+     */
+    public Builder withServices(Class<? extends ServiceModule> serviceModule,
+                                Class<? extends ServiceModule>... serviceModules) {
+      List<Class<? extends ServiceModule>> services = asList(serviceModule, serviceModules);
+      this.services.addAll(services);
+      return this;
+    }
+
+    /**
+     * Adds a list of services with which the TestKit would be instantiated.
+     */
+    public Builder withServices(Iterable<Class<? extends ServiceModule>> serviceModules) {
+      serviceModules.forEach(services::add);
```

Comment:
I find this one a little clearer (addAll instead of generic for each):
```suggestion
      Iterables.addAll(services, serviceModules);
```